src/docfx/build/toc/parser/MarkdownTocMarkup.cs

```diff
@@ -0,0 +1,210 @@
+        {
+            if (headingBlocks.Length == 0)
+            {
+                return (new List<TableOfContentsInputItem>(), 0);
+            }
+
+            Debug.Assert(startIndex < headingBlocks.Length);
+
+            int i = startIndex;
+            var items = new List<TableOfContentsInputItem>();
+            var childrenCount = 0;
+            do
+            {
+                var item = GetItem(headingBlocks[i]);
+                if (item != null)
+                {
+                    items.Add(item);
+                    var currentLevel = headingBlocks[i].Level;
+                    if (i + 1 < headingBlocks.Length && headingBlocks[i + 1].Level > currentLevel)
+                    {
+                        if (headingBlocks[i + 1].Level - currentLevel > 1)
+                        {
+                            throw Errors.InvalidTocLevel(filePath, currentLevel, headingBlocks[i + 1].Level).ToException();
+                        }
+
+                        var (children, count) = ConvertTo(filePath, headingBlocks, i + 1);
+                        item.Items = children;
+                        i += count;
+                        childrenCount += count;
+                    }
+
+                    if (i + 1 < headingBlocks.Length && headingBlocks[i + 1].Level < currentLevel)
+                    {
+                        break;
+                    }
+                }
+            }
+            while (++i < headingBlocks.Length);
+
+            return (items, items.Count + childrenCount);
+
+            TableOfContentsInputItem GetItem(HeadingBlock block)
+            {
+                if (block.Inline == null || !block.Inline.Any())
+                {
+                    return null;
+                }
+
+                string name = null;
+                string displayName = null;
+                string href = null;
+
+                if (block.Inline.First() is LinkInline linkInline)
+                {
+                    if (!string.IsNullOrEmpty(linkInline.Url))
+                        href = linkInline.Url;
+                    if (!string.IsNullOrEmpty(linkInline.Title))
+                        displayName = linkInline.Title;
+
+                    if (linkInline.Any() && linkInline.First() is LiteralInline literalInline)
+                    {
+                        name = literalInline.Content.ToString();
+                    }
+                }
+                else if (block.Inline.First() is LiteralInline literalInline)
+                {
+                    name = literalInline.Content.ToString();
+                }
+
+                var currentItem = new TableOfContentsInputItem
+                {
+                    DisplayName = displayName,
+                    Name = name,
+                    Href = href,
+                };
+
+                return currentItem;
+            }
+        }
+
+        private static MarkdownPipeline CreateMarkdownPipeline()
```

Comment:
>CreateMarkdownPipeline [](start = 40, length = 22)

add `(errors, ast) Markup.Parse(string markdown, Document file)` method to avoid the duplications below